clientconn.go

```diff
@@ -788,6 +790,87 @@ func (cc *ClientConn) getTransport(ctx context.Context, failfast bool) (transpor
 	return t, done, nil
 }
 
+func parseTimeout(t *string) (*time.Duration, error) {
+	if t == nil {
+		return nil, nil
+	}
+	d, err := time.ParseDuration(*t)
+	return &d, err
+}
+
+type jsonName struct {
+	Service *string `json:"service,omitempty"`
+	Method  *string `json:"method,omitempty"`
+}
+
+func (j jsonName) Stringer() string {
+	res := ""
+	if j.Service != nil {
+		res += "/" + *j.Service + "/"
+	}
+	if j.Method != nil {
+		res += *j.Method
+	}
+	return res
+}
+
+type jsonMC struct {
+	Name                    *[]jsonName `json:"name,omitempty"`
+	WaitForReady            *bool       `json:"waitForReady,omitempty"`
+	Timeout                 *string     `json:"timeout,omitempty"`
+	MaxRequestMessageBytes  *int        `json:"maxRequestMessageBytes,omitempty"`
+	MaxResponseMessageBytes *int        `json:"maxResponseMessageBytes,omitempty"`
+}
+
+type jsonSC struct {
+	LoadBalancingPolicy *string   `json:"loadBalancingPolicy,omitempty"`
+	MethodConfig        *[]jsonMC `json:"methodConfig,omitempty"`
+}
+
+// handleServiceConfig parses the service config string in JSON format to Go native
+// struct ServiceConfig, and store both the struct and the JSON string in ClientConn.
+func (cc *ClientConn) handleServiceConfig(js string) error {
```

Comment:
Move the all service config related code (structs and functions) to a new file `service_config.go` and also add unit tests for the json parsing function.

Put all the parsing code in a function
```go
func parseJsonServiceConfig(js string) (ServiceConfig, error)
```

In `ClientConn`, you can do
```go
if sc, err := parse(js); err == nil {
  cc.mu.Lock()
  cc.scRaw = js
  cc.sc = sc
  cc.mu.Unlock()
}
```

Reply:
done